### PR TITLE
MAINT: Remove any promotion-state switching logic

### PR DESCRIPTION
**Pull request by seberg**
This is the first level, no following code simplifications, just straight up deletions of any branching.

I kept a UserWarning, in case someone had the bad idea to permanently set the environment variable and think they can rely on it. Although would be happy to just delete that as well.

---

I don't really think this interferes with regularly touched code, so think we might as well do it now and not worry about backports (the diff isn't huge anyway)

### CODE REVIEW

File: numpy/_core/tests/test_numeric.py [hunk 1]
```diff
@@ -1488,21 +1488,22 @@ def test_can_cast_structured_to_simple(self):
...
+        # that was approximately whether it would overflow to inf).
+        with pytest.raises(TypeError):
+            np.can_cast(4, "int8", casting="unsafe")
+
+        with pytest.raises(TypeError):
+            np.can_cast(4.0, "float64", casting="unsafe")
+
+        with pytest.raises(TypeError):
+            np.can_cast(4j, "complex128", casting="unsafe")
+
```
Comment (ngoldbaum) [inline numpy/_core/tests/test_numeric.py:1506]: Not immediately obvious that the test is testing the same thing with the changes, although I see it was xfailing before so this is probably more useful.

**Review by ngoldbaum — APPROVED** [review]
Always love deleting almost a thousand lines of sort of complicated C code. I left a few minor comments but this looks good.

File: numpy/_core/tests/test_numeric.py (hunk 1 above)
Comment (seberg) [reply numpy/_core/tests/test_numeric.py:1506]: The old test checked the point where value-based casting flipped.  That doesn't make sense if we don't inspect the value...

File: numpy/_core/src/multiarray/multiarraymodule.c [hunk 2]
```diff
@@ -4627,14 +4615,6 @@ static struct PyMethodDef array_module_methods[] = {
...
         (PyCFunction) get_handler_version,
         METH_VARARGS, NULL},
-    {"_get_promotion_state",
-        (PyCFunction)npy__get_promotion_state,
-        METH_NOARGS, "Get the current NEP 50 promotion state."},
-    {"_set_promotion_state",
-         (PyCFunction)npy__set_promotion_state,
-         METH_O, "Set the NEP 50 promotion state.  This is not thread-safe.\n"
-                 "The optional warnings can be safely silenced using the \n"
-                 "`np._no_nep50_warning()` context manager."},
```
[the remark this answers is not in the document]
Comment (seberg) [reply numpy/_core/src/multiarray/multiarraymodule.c:4637]: Added a bit for that and also some other doc tweaks.  Not perfect, but maybe a decent start.  would be nice if you can have a glance over it, though.

**Review by mattip — APPROVED** [review]
LGTM, and the documentation changes make sense. I will leave this for a bit if anyone else wants to take a look.

File: numpy/_core/tests/test_dtype.py [hunk 3]
```diff
@@ -1433,34 +1433,26 @@ class TestPromotion:
...
-             (np.clongdouble(np.nextafter(1.7e308, np.inf)),
-                  np.clongdouble, None),
+             (np.clongdouble(np.nextafter(1.7e308, np.inf)), np.clongdouble),
              ])
     def test_complex_other_value_based(self,
-            weak_promotion, other, expected, expected_weak):
-        if weak_promotion and expected_weak is not None:
-            expected = expected_weak
-
+            other, expected):
```
Comment (mhvk) [inline numpy/_core/tests/test_dtype.py:1455]: Bring back to one line?

**Review by mhvk — APPROVED** [review]
Looks good, but there are quite a few comments by @ngoldbaum that are unresolved (though most/all are addressed?), so just approving, not merging.

Also one tiny nitpick - not worth running CI on its own, but can be done if there are other changes to be made still.

**Comment by ngoldbaum** [comment]
Sorry for not giving this another pass! Looks good to me now. @seberg feel free to merge along with Marten's suggested fixup.

File: numpy/__init__.py [hunk 4]
```diff
@@ -529,8 +529,11 @@ def hugepage_setup():
     _core.multiarray._multiarray_umath._reload_guard()
 
     # TODO: Remove the environment variable entirely now that it is "weak"
-    _core._set_promotion_state(
-        os.environ.get("NPY_PROMOTION_STATE", "weak"))
+    if (os.environ.get("NPY_PROMOTION_STATE", "weak") != "weak"):
```
Comment (akaashhazarika) [inline numpy/__init__.py:532]: Assign "weak" to a variable like DEFAULT_PROMOTION_STATE for better maintainability and readability.

**Comment by charris** [comment]
@seberg I put up a backport for this (#27397) in order to simplify some other backports.

**Comment by seberg** [comment]
Hmmm, this removes the ability to use the env variable, so not sure about back-porting.

**Comment by charris** [comment]
@seberrg OK, I'll skip a backport for now.

**Pull request merged**